File: web/src/features/jobs/chat/mentionToken.ts
/// mention ฝังเป็น token ในข้อความเอง รูปแบบ "@[staffId:ชื่อ]" — ไม่ parse จาก plain text ฝั่ง server
/// (server รับ mentionedStaffIds แยกมาให้ validate ตรงๆ) client เป็นคนสร้าง/ตัด render เอง
/// ชื่อพนักงานถูก sanitize ตัดอักขระ [ ] : ออกก่อนฝัง กันชนกับ delimiter ของ token เอง

export type ChatBodySegment =
  | { type: 'text'; value: string }
  | { type: 'mention'; staffId: number; name: string }

const MENTION_PATTERN = /@\[(\d+):([^[\]]+)]/g

export function sanitizeMentionName(name: string): string {
  return name.replace(/[[\]:]/g, '').trim()
}

export function buildMentionToken(staffId: number, name: string): string {
  return `@[${staffId}:${sanitizeMentionName(name)}]`
}

export function parseChatBody(body: string): ChatBodySegment[] {
  const segments: ChatBodySegment[] = []
  let lastIndex = 0
  MENTION_PATTERN.lastIndex = 0
  let match: RegExpExecArray | null
  // eslint-disable-next-line no-cond-assign
  while ((match = MENTION_PATTERN.exec(body)) !== null) {
    if (match.index > lastIndex) segments.push({ type: 'text', value: body.slice(lastIndex, match.index) })
    segments.push({ type: 'mention', staffId: Number(match[1]), name: match[2] ?? '' })
    lastIndex = match.index + (match[0]?.length ?? 0)
  }
  if (lastIndex < body.length) segments.push({ type: 'text', value: body.slice(lastIndex) })
  return segments
}

export function extractMentionedStaffIds(body: string): number[] {
  const ids = parseChatBody(body)
    .filter((segment): segment is Extract<ChatBodySegment, { type: 'mention' }> => segment.type === 'mention')
    .map((segment) => segment.staffId)
  return Array.from(new Set(ids))
}

/// หาตำแหน่ง "@" ที่กำลังพิมพ์ค้นหาอยู่ก่อนตำแหน่ง caret (ไม่มีช่องว่าง/ไม่อยู่หลัง token ที่ปิดแล้ว)
/// คืน null ถ้าไม่ได้อยู่ระหว่างพิมพ์ mention
export function findMentionTriggerIndex(value: string, caret: number): number | null {
  for (let i = caret - 1; i >= 0; i--) {
    const ch = value[i]
    if (ch === '@') return i
    if (ch === undefined || /\s/.test(ch) || ch === ']') return null
  }
  return null
}

File: mobile/lib/features/jobs/chat/mention_token.dart
/// mention ฝังเป็น token ในข้อความเอง รูปแบบ "@[staffId:ชื่อ]"
///
/// **ต้องให้ผลตรงกับ web/src/features/jobs/chat/mentionToken.ts ทุกกรณี**
/// ถ้าสองฝั่ง parse ไม่ตรงกัน อีกฝั่งจะเห็นข้อความดิบ "@[41:สมชาย]" แทนชื่อ
/// server ไม่ parse ข้อความ — รับ mentionedStaffIds แยกมา validate ตรงๆ
library;

sealed class ChatBodySegment {
  const ChatBodySegment();
}

class TextSegment extends ChatBodySegment {
  const TextSegment(this.value);
  final String value;
}

class MentionSegment extends ChatBodySegment {
  const MentionSegment(this.staffId, this.name);
  final int staffId;
  final String name;
}

final _mentionPattern = RegExp(r'@\[(\d+):([^\[\]]+)\]');

/// ตัด [ ] : ออกจากชื่อก่อนฝัง กันชนกับ delimiter ของ token เอง
String sanitizeMentionName(String name) => name.replaceAll(RegExp(r'[\[\]:]'), '').trim();

String buildMentionToken(int staffId, String name) => '@[$staffId:${sanitizeMentionName(name)}]';

List<ChatBodySegment> parseChatBody(String body) {
  final segments = <ChatBodySegment>[];
  var lastIndex = 0;

  for (final match in _mentionPattern.allMatches(body)) {
    if (match.start > lastIndex) {
      segments.add(TextSegment(body.substring(lastIndex, match.start)));
    }
    segments.add(MentionSegment(int.parse(match.group(1)!), match.group(2) ?? ''));
    lastIndex = match.end;
  }

  if (lastIndex < body.length) segments.add(TextSegment(body.substring(lastIndex)));
  return segments;
}

List<int> extractMentionedStaffIds(String body) {
  final ids = <int>{};
  for (final segment in parseChatBody(body)) {
    if (segment is MentionSegment) ids.add(segment.staffId);
  }
  return ids.toList();
}

/// หาตำแหน่ง "@" ที่กำลังพิมพ์ค้นหาอยู่ก่อน caret — คืน null ถ้าไม่ได้อยู่ระหว่างพิมพ์ mention
int? findMentionTriggerIndex(String value, int caret) {
  if (caret < 0 || caret > value.length) return null;

  for (var i = caret - 1; i >= 0; i--) {
    final ch = value[i];
    if (ch == '@') return i;
    if (ch == ']' || RegExp(r'\s').hasMatch(ch)) return null;
  }
  return null;
}
